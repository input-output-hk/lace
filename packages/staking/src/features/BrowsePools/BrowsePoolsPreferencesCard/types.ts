import { PoolsFilter } from '../../store';

export type SelectOption = { label: string; value: string; selected: boolean };

export type FilterOption = {
  key: PoolsFilter;
  title: string;
  type: 'input' | 'select';
  opts: string[] | SelectOption[];
};

export enum SortAndFilterTab {
  sort = 'sort',
  filter = 'filter',
}
