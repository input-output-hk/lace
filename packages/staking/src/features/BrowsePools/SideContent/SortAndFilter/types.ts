export enum PoolsFilter {
  Saturation = 'SATURATION',
  ProfitMargin = 'PROFIT_MARGIN',
  Performance = 'PERFORMANCE',
  Ros = 'ROS',
}

export type SortOption = { label: string; value: string; icon: any; onIconClick: any };
export type SelectOption = { label: string; value: string; selected: boolean };

export type FilterOption = {
  key: PoolsFilter;
  title: string;
  type: 'input' | 'select';
  opts: string[] | SelectOption[];
};

export interface FilterValues {
  [PoolsFilter.Saturation]: string[];
  [PoolsFilter.ProfitMargin]: string[];
  [PoolsFilter.Performance]: string[];
  [PoolsFilter.Ros]: string[];
}

export type VisibleSection = 'sorting' | 'filtering';
