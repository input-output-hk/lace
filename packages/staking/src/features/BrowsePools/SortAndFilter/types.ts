export enum PoolsFilter {
  Saturation = 'SATURATION',
  ProfitMargin = 'PROFIT_MARGIN',
  Performance = 'PERFORMANCE',
  Ros = 'ROS',
}

export type SelectOption = { label: string; value: string; selected: boolean };

export type FilterOption = {
  key: PoolsFilter;
  title: string;
  type: 'input' | 'select';
  opts: string[] | SelectOption[];
};

export interface FilterValues {
  [PoolsFilter.Saturation]: string[];
  [PoolsFilter.ProfitMargin]: string[];
  [PoolsFilter.Performance]: string[];
  [PoolsFilter.Ros]: string[];
}

export enum SortAndFilterTab {
  sort = 'sort',
  filter = 'filter',
}
