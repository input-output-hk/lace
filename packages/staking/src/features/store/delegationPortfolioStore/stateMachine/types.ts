import { Wallet } from '@lace/cardano';
import { AdaSymbol } from '../types';
import { Command } from './commands';

export type StakePoolWithLogo = Wallet.Cardano.StakePool & { logo?: string };

type DraftPortfolioStakePoolBase = {
  id: Wallet.Cardano.Cip17Pool['id'];
  targetWeight: Wallet.Cardano.Cip17Pool['weight'];
  displayData: Wallet.util.StakePool;
  stakePool: Wallet.Cardano.StakePool;
};

export type DraftPortfolioStakePool = DraftPortfolioStakePoolBase &
  (
    | {
        basedOnCurrentPortfolio: false;
      }
    | { basedOnCurrentPortfolio: true; currentPortfolioPercentage: Wallet.Percent }
  );

export type CurrentPortfolioStakePool = DraftPortfolioStakePoolBase & {
  displayData: Wallet.util.StakePool & {
    lastReward: bigint;
    totalRewards: bigint;
  };
  value: bigint;
  percentage: Wallet.Percent;
};

export enum Flow {
  Overview = 'Overview',
  BrowsePools = 'BrowsePools',
  CurrentPoolDetails = 'CurrentPoolDetails',
  PoolDetails = 'PoolDetails',
  PortfolioManagement = 'PortfolioManagement',
  ChangingPreferences = 'ChangingPreferences',
  NewPortfolio = 'NewPortfolio',
}

export type ExpandedViewFlow = Flow;

export type PopupViewFLow = Flow.Overview | Flow.CurrentPoolDetails;

export enum DrawerDefaultStep {
  PoolDetails = 'PoolDetails',
}

export enum DrawerManagementStep {
  Preferences = 'Preferences',
  Confirmation = 'Confirmation',
  Sign = 'Sign',
  Success = 'Success',
  Failure = 'Failure',
}

export type DrawerStep = DrawerDefaultStep | DrawerManagementStep;

export type State = {
  activeDrawerStep?: DrawerStep;
  activeFlow: Flow;
  cardanoCoinSymbol: AdaSymbol;
  pendingSelectedPortfolio?: DraftPortfolioStakePool[];
  currentPortfolio: CurrentPortfolioStakePool[];
  draftPortfolio?: DraftPortfolioStakePool[];
  selectedPortfolio: DraftPortfolioStakePool[];
  viewedStakePool?: StakePoolWithLogo;
};

export type ExecuteCommand = <C extends Command>(command: C) => void;

export type Handler<C extends Command = any> = (params: {
  command: C;
  executeCommand: ExecuteCommand;
  state: State;
}) => void;
