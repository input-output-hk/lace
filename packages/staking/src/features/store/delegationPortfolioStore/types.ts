import { DelegatedStake } from '@cardano-sdk/wallet';
import { Wallet } from '@lace/cardano';
import { ExecuteCommand, State } from './stateMachine';

export type AdaSymbol = 'ADA' | 'tADA';

export type DelegationPortfolioState = State & {
  view?: 'popup' | 'expanded';
};

export type DelegationPortfolioStore = DelegationPortfolioState & {
  mutators: {
    executeCommand: ExecuteCommand;
    forceAbortFlows: () => void;
    setCardanoCoinSymbol: (currentChain: Wallet.Cardano.ChainId) => void;
    setCurrentPortfolio: (params: {
      delegationDistribution: DelegatedStake[];
      currentEpoch: Wallet.EpochInfo;
      delegationRewardsHistory: Wallet.RewardsHistory;
    }) => Promise<void>;
    setView: (view: 'popup' | 'expanded') => void;
  };
};

export type StakePoolDetails = {
  delegators?: number | string;
  description: string;
  hexId: string;
  id: string;
  logo?: string;
  margin: number | string;
  name: string;
  owners: string[];
  saturation?: number | string;
  stake: { number: string; unit?: string };
  ticker: string;
  apy?: number | string;
  status: Wallet.Cardano.StakePool['status'];
  fee: number | string;
  contact: Wallet.Cardano.PoolContactData;
};
