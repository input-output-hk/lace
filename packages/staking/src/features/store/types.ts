import { DelegatedStake } from '@cardano-sdk/wallet';
import { Wallet } from '@lace/cardano';

export enum Sections {
  DETAIL = 'detail',
  PREFERENCES = 'preferences',
  CONFIRMATION = 'confirmation',
  SIGN = 'sign',
  SUCCESS_TX = 'success_tx',
  FAIL_TX = 'fail_tx',
}

export enum StakingError {
  UTXO_FULLY_DEPLETED = 'UTXO_FULLY_DEPLETED',
  UTXO_BALANCE_INSUFFICIENT = 'UTXO_BALANCE_INSUFFICIENT',
}

export interface SectionConfig {
  currentSection: Sections;
  nextSection?: Sections;
  prevSection?: Sections;
}

export interface StakePoolDetails {
  simpleSendConfig: SectionConfig;
  setSection: (section?: SectionConfig) => void;
  activePage: Page;
  setActivePage: (page: Page) => void;
  setPrevSection: () => void;
  resetStates: () => void;
  isDrawerVisible: boolean;
  setIsDrawerVisible: (visibility: boolean) => void;
  isStakeConfirmationVisible: boolean;
  setStakeConfirmationVisible: (visibility: boolean) => void;
  isExitStakingVisible: boolean;
  setExitStakingVisible: (visibility: boolean) => void;
  isNoFundsVisible: boolean;
  setNoFundsVisible: (visibility: boolean) => void;
  isBuildingTx: boolean;
  setIsBuildingTx: (visibility: boolean) => void;
  stakingError?: StakingError;
  setStakingError: (error?: StakingError) => void;
}

export type DraftPortfolioStakePool = Wallet.Cardano.Cip17Pool & {
  displayData: Wallet.util.StakePool;
};

export type CurrentPortfolioStakePool = DraftPortfolioStakePool & {
  stakePool: Wallet.Cardano.StakePool;
  value: bigint;
};

export type DelegationPortfolioState = {
  currentPortfolio: CurrentPortfolioStakePool[];
  draftPortfolio: DraftPortfolioStakePool[];
};

export type DelegationPortfolioQueries = {
  poolIncludedInDraft: (id: Wallet.Cardano.PoolIdHex) => boolean;
};

type DelegationPortfolioMutators = {
  setCurrentPortfolio: (params: {
    delegationDistribution: DelegatedStake[];
    cardanoCoin: Wallet.CoinId;
  }) => Promise<void>;
  addPoolToDraft: (pool: DraftPortfolioStakePool) => void;
  removePoolFromDraft: (params: Pick<DraftPortfolioStakePool, 'id'>) => void;
  updatePoolWeight: (params: Pick<DraftPortfolioStakePool, 'id' | 'weight'>) => void;
  clearDraft: () => void;
};

export type DelegationPortfolioStore = DelegationPortfolioState &
  DelegationPortfolioQueries & {
    mutators: DelegationPortfolioMutators;
  };

export enum Page {
  overview = 'overview',
  browsePools = 'browsePools',
}
