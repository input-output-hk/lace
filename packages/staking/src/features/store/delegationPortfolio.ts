import { Wallet } from '@lace/cardano';
import { create } from 'zustand';
import { immer } from 'zustand/middleware/immer';
import { DelegationPortfolioState, DelegationPortfolioStore, PortfolioManagementProcess } from './types';

const defaultState: DelegationPortfolioState = {
  activeManagementProcess: PortfolioManagementProcess.None,
  currentPortfolio: [],
  draftPortfolio: [],
  selections: [],
};

export const MAX_POOLS_COUNT = 5;

export const useDelegationPortfolioStore = create(
  immer<DelegationPortfolioStore>((set, get) => ({
    ...defaultState,
    mutators: {
      beginManagementProcess: (process) =>
        set((store) => {
          if (store.activeManagementProcess === process) return;
          store.activeManagementProcess = process;
          store.draftPortfolio =
            process === PortfolioManagementProcess.CurrentPortfolio ? store.currentPortfolio : store.selections;
        }),
      // eslint-disable-next-line unicorn/no-object-as-default-parameter
      cancelManagementProcess: ({ dumpDraftToSelections } = { dumpDraftToSelections: false }) =>
        set((store) => {
          if (store.activeManagementProcess === PortfolioManagementProcess.None) return;
          if (dumpDraftToSelections) {
            store.selections = store.draftPortfolio.map((pool) => ({
              ...pool,
              weight: 1,
            }));
          }
          store.draftPortfolio = [];
          store.activeManagementProcess = PortfolioManagementProcess.None;
        }),
      clearSelections: () =>
        set((store) => {
          store.selections = [];
        }),
      finalizeManagementProcess: () =>
        set((store) => {
          if (store.activeManagementProcess === PortfolioManagementProcess.None) return;
          store.draftPortfolio = [];
          if (store.activeManagementProcess === PortfolioManagementProcess.NewPortfolio) {
            store.selections = [];
          }
          store.activeManagementProcess = PortfolioManagementProcess.None;
        }),
      removePoolInManagementProcess: ({ id }) =>
        set((store) => {
          if (store.activeManagementProcess === PortfolioManagementProcess.None) return;
          store.draftPortfolio = store.draftPortfolio.filter((pool) => pool.id !== id);
          if (store.activeManagementProcess === PortfolioManagementProcess.NewPortfolio) {
            store.selections = store.draftPortfolio;
          }
        }),
      selectPool: (poolData) =>
        set(({ selections }) => {
          const { selectionsFull } = get().queries;
          const alreadySelected = selections.some(({ id }) => poolData.id === id);
          if (selectionsFull() || alreadySelected) return;
          selections.push(poolData);
        }),
      setCurrentPortfolio: async ({ cardanoCoin, delegationDistribution }) => {
        const currentPortfolio = delegationDistribution.map(({ pool: stakePool, percentage, stake }) => ({
          displayData: Wallet.util.stakePoolTransformer({ cardanoCoin, stakePool }),
          id: stakePool.hexId,
          name: stakePool.metadata?.name,
          stakePool,
          ticker: stakePool.metadata?.ticker,
          value: stake,
          weight: percentage,
        }));

        set((store) => {
          store.currentPortfolio = currentPortfolio;
        });
      },
      unselectPool: ({ id }) =>
        set((store) => {
          store.selections = store.selections.filter((pool) => pool.id !== id);
        }),
    },
    queries: {
      isPoolSelected: (hexId) => {
        const { selections } = get();
        return !!selections?.find((pool) => pool.id === hexId);
      },
      selectionsFull: () => {
        const { selections } = get();
        return selections.length === MAX_POOLS_COUNT;
      },
    },
  }))
);
