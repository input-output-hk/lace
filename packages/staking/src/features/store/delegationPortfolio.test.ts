/* eslint-disable no-magic-numbers, sonarjs/no-identical-functions */
import { Wallet } from '@lace/cardano';
import { act, renderHook } from '@testing-library/react-hooks';
import { beforeEach, expect, it } from 'vitest';
import { MAX_POOLS_COUNT, useDelegationPortfolioStore } from './delegationPortfolio';
import { CurrentPortfolioStakePool, PortfolioManagementProcess } from './types';

const dummyPool1 = {
  id: Wallet.Cardano.PoolIdHex('39deffa1dfcfe192ea0efeb3e9bcd9878190627fb590ec81f390cd6d'),
  name: '8BOOL',
  ticker: '8BOOL',
  weight: 1,
} as CurrentPortfolioStakePool;
const dummyPool2 = {
  id: Wallet.Cardano.PoolIdHex('39faf89aba3daab2bce656a9698b0a2c05e679c0fb360ee6e3b46acb'),
  name: '8BOOM',
  ticker: '8BOOM',
  weight: 1,
} as CurrentPortfolioStakePool;
const dummyPool3 = {
  id: Wallet.Cardano.PoolIdHex('3867a09729a1f954762eea035a82e2d9d3a14f1fa791a022ef0da242'),
  name: 'ADA Capital',
  ticker: 'ADACT',
  weight: 1,
} as CurrentPortfolioStakePool;
const dummyPool4 = {
  id: Wallet.Cardano.PoolIdHex('a0e79024226e4febf20214164d88dcd269c54819fc3b810ca5cc45a5'),
  name: 'Example Pool 4',
  weight: 1,
} as CurrentPortfolioStakePool;
const dummyPool5 = {
  id: Wallet.Cardano.PoolIdHex('cde2511b7638ab734db8534daebb9a2243c4ef1694c82f85b8825be9'),
  name: 'Example Pool 5',
  weight: 1,
} as CurrentPortfolioStakePool;
const dummyPool6 = {
  id: Wallet.Cardano.PoolIdHex('b546d6339727ae557830265c581381735a4f797591ff8f56e14082c6'),
  name: 'Example Pool 6',
  weight: 1,
} as CurrentPortfolioStakePool;

const dummyStakePool1 = {
  cost: BigInt(340_000_000),
  hexId: dummyPool1.id,
  id: Wallet.Cardano.PoolId('pool18800lgwlelse96swl6e7n0xes7qeqcnlkkgweq0njrxk66zlj9a'),
  metadata: {
    name: '8BOOL',
    ticker: '8BOOL',
  },
  metrics: {},
  pledge: BigInt(1_000_000_000),
} as Wallet.Cardano.StakePool;
const dummyStakePool2 = {
  cost: BigInt(340_000_000),
  hexId: dummyPool2.id,
  id: Wallet.Cardano.PoolId('pool188a03x468k4t908x265knzc29sz7v7wqlvmqaehrk34vkll3e2h'),
  metadata: {
    name: '8BOOM',
    ticker: '8BOOM',
  },
  metrics: {},
  pledge: BigInt(7_000_000_000),
} as Wallet.Cardano.StakePool;
const dummyStakePool3 = {
  cost: BigInt(340_000_000),
  hexId: dummyPool3.id,
  id: Wallet.Cardano.PoolId('pool18pn6p9ef58u4ga3wagp44qhzm8f6zncl57g6qgh0pk3yytwz54h'),
  metadata: {
    name: 'ADA Capital',
    ticker: 'ADACT',
  },
  metrics: {},
  pledge: BigInt(1_659_000_000_000),
} as Wallet.Cardano.StakePool;

describe('delegationPortfolioStore', () => {
  it('initializes the portfolio preferences', () => {
    const { result } = renderHook(() => useDelegationPortfolioStore());
    expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.None);
    expect(result.current.currentPortfolio).toEqual([]);
    expect(result.current.draftPortfolio).toEqual([]);
    expect(result.current.selections).toEqual([]);
  });

  it('sets the current portfolio', () => {
    const expectedLength = 3;
    const { result } = renderHook(() => useDelegationPortfolioStore());
    act(() =>
      result.current.mutators.setCurrentPortfolio({
        cardanoCoin: { symbol: 'ADA' } as Wallet.CoinId,
        delegationDistribution: [
          {
            percentage: Wallet.Percent(33.33),
            pool: dummyStakePool1,
            rewardAccounts: [],
            stake: BigInt(1),
          },
          {
            percentage: Wallet.Percent(33.33),
            pool: dummyStakePool2,
            rewardAccounts: [],
            stake: BigInt(1),
          },
          {
            percentage: Wallet.Percent(33.33),
            pool: dummyStakePool3,
            rewardAccounts: [],
            stake: BigInt(1),
          },
        ],
      })
    );
    expect(result.current.currentPortfolio.length).toEqual(expectedLength);
    expect(result.current.currentPortfolio).toEqual([
      expect.objectContaining({
        ...dummyPool1,
        weight: 33.33,
      }),
      expect.objectContaining({
        ...dummyPool2,
        weight: 33.33,
      }),
      expect.objectContaining({
        ...dummyPool3,
        weight: 33.33,
      }),
    ]);
  });

  describe('selections', () => {
    beforeEach(() => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.clearSelections());
    });

    it('allows for selecting pools preserving the max allowed count', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.selectPool(dummyPool1));
      act(() => result.current.mutators.selectPool(dummyPool2));
      act(() => result.current.mutators.selectPool(dummyPool3));
      act(() => result.current.mutators.selectPool(dummyPool4));
      act(() => result.current.mutators.selectPool(dummyPool5));
      act(() => result.current.mutators.selectPool(dummyPool6));
      expect(result.current.selections.length).toEqual(MAX_POOLS_COUNT);
    });

    it('prevents selecting a same pool twice', () => {
      const newLength = 1;
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.selectPool(dummyPool1));
      act(() => result.current.mutators.selectPool(dummyPool1));
      expect(result.current.selections.length).toEqual(newLength);
    });

    it('allows to unselect pool', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.selectPool(dummyPool1));
      act(() => result.current.mutators.selectPool(dummyPool2));
      act(() => result.current.mutators.unselectPool({ id: dummyPool1.id }));
      act(() => result.current.mutators.unselectPool({ id: dummyPool3.id }));
      expect(result.current.selections.length).toEqual(1);
      expect(result.current.selections[0]?.id).toEqual(dummyPool2.id);
    });

    it('clears the selections', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.selectPool(dummyPool1));
      act(() => result.current.mutators.selectPool(dummyPool2));
      act(() => result.current.mutators.clearSelections());
      expect(result.current.selections.length).toEqual(0);
    });
  });

  describe('current portfolio management', () => {
    beforeEach(() => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.setCurrentPortfolio({
          cardanoCoin: { symbol: 'ADA' } as Wallet.CoinId,
          delegationDistribution: [
            {
              percentage: Wallet.Percent(33.33),
              pool: dummyStakePool1,
              rewardAccounts: [],
              stake: BigInt(1),
            },
            {
              percentage: Wallet.Percent(33.33),
              pool: dummyStakePool2,
              rewardAccounts: [],
              stake: BigInt(1),
            },
            {
              percentage: Wallet.Percent(33.33),
              pool: dummyStakePool3,
              rewardAccounts: [],
              stake: BigInt(1),
            },
          ],
        });
      });
      act(() => result.current.mutators.beginManagementProcess(PortfolioManagementProcess.CurrentPortfolio));
    });

    it('sets the currentPortfolio as activeManagementProcess when started', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.CurrentPortfolio);
    });
    it('copies the currentPortfolio items to the draftPortfolio when started', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      expect(result.current.draftPortfolio).toEqual(result.current.currentPortfolio);
    });
    it('re-sets the activeManagementProcess to none when canceled', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.cancelManagementProcess();
      });
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.None);
    });
    it('clears the draftPortfolio when canceled', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.cancelManagementProcess();
      });
      expect(result.current.draftPortfolio).toEqual([]);
    });
    it('dumps draftPortfolio to selections when canceled with dumpDraftToSelections flag', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.cancelManagementProcess({ dumpDraftToSelections: true });
      });
      expect(result.current.selections).toEqual([
        expect.objectContaining(dummyPool1),
        expect.objectContaining(dummyPool2),
        expect.objectContaining(dummyPool3),
      ]);
    });
    it('re-sets the activeManagementProcess to none when finalized', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.finalizeManagementProcess();
      });
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.None);
    });
    it('clears the draftPortfolio when finalized', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.finalizeManagementProcess();
      });
      expect(result.current.draftPortfolio).toEqual([]);
    });
    it('allows to remove pool from draftPortfolio', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.removePoolInManagementProcess({ id: dummyPool1.id }));
      expect(result.current.draftPortfolio).toEqual([
        expect.objectContaining({
          ...dummyPool2,
          weight: 33.33,
        }),
        expect.objectContaining({
          ...dummyPool3,
          weight: 33.33,
        }),
      ]);
    });
  });

  describe('new portfolio management', () => {
    beforeEach(() => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.selectPool(dummyPool2));
      act(() => result.current.mutators.selectPool(dummyPool3));
      act(() => result.current.mutators.selectPool(dummyPool4));
      act(() => result.current.mutators.selectPool(dummyPool5));
      act(() => result.current.mutators.beginManagementProcess(PortfolioManagementProcess.NewPortfolio));
    });

    it('sets the newPortfolio as activeManagementProcess when started', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.NewPortfolio);
    });
    it('copies the selections items to the draftPortfolio when started', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      expect(result.current.draftPortfolio).toEqual(result.current.selections);
    });
    it('re-sets the activeManagementProcess to none when canceled', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.cancelManagementProcess();
      });
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.None);
    });
    it('clears the draftPortfolio when canceled', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.cancelManagementProcess();
      });
      expect(result.current.draftPortfolio).toEqual([]);
    });
    it('re-sets the activeManagementProcess to none when finalized', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.finalizeManagementProcess();
      });
      expect(result.current.activeManagementProcess).toEqual(PortfolioManagementProcess.None);
    });
    it('clears the draftPortfolio when finalized', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.finalizeManagementProcess();
      });
      expect(result.current.draftPortfolio).toEqual([]);
    });
    it('clears the selections when finalized', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => {
        result.current.mutators.finalizeManagementProcess();
      });
      expect(result.current.selections).toEqual([]);
    });
    it('allows to remove pool from draftPortfolio', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.removePoolInManagementProcess({ id: dummyPool3.id }));
      expect(result.current.draftPortfolio).toEqual([dummyPool2, dummyPool4, dummyPool5]);
    });
    it('removes pool from selections when removed pool in management process', () => {
      const { result } = renderHook(() => useDelegationPortfolioStore());
      act(() => result.current.mutators.removePoolInManagementProcess({ id: dummyPool1.id }));
      expect(result.current.selections).toEqual([dummyPool2, dummyPool4, dummyPool5]);
    });
  });
});
