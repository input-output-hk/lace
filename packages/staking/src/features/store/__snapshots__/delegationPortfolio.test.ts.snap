// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`delegationPortfolioStore > sets the current portfolio 1`] = `
[
  {
    "displayData": {
      "apy": undefined,
      "cost": "-% + 340ADA",
      "description": undefined,
      "fee": "340.00",
      "hexId": "39deffa1dfcfe192ea0efeb3e9bcd9878190627fb590ec81f390cd6d",
      "id": "pool18800lgwlelse96swl6e7n0xes7qeqcnlkkgweq0njrxk66zlj9a",
      "isStakingPool": undefined,
      "lastReward": 1111n,
      "logo": "data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2230%22%20height%3D%2230%22%20viewBox%3D%220%200%2030%2030%22%3E%3Cpath%20fill%3D%22%23e5e5e5%22%20d%3D%22M15%209L9%209L9%206ZM15%209L15%203L18%203ZM15%2021L21%2021L21%2024ZM15%2021L15%2027L12%2027ZM9%2015L3%2015L3%2012ZM21%2015L21%209L24%209ZM21%2015L27%2015L27%2018ZM9%2015L9%2021L6%2021Z%22%2F%3E%3Cpath%20fill%3D%22%238bcc66%22%20d%3D%22M3%206L6%203L9%206L6%209ZM24%203L27%206L24%209L21%206ZM27%2024L24%2027L21%2024L24%2021ZM6%2027L3%2024L6%2021L9%2024ZM15%2012L15%2015L12%2015ZM18%2015L15%2015L15%2012ZM15%2018L15%2015L18%2015ZM12%2015L15%2015L15%2018Z%22%2F%3E%3C%2Fsvg%3E",
      "margin": "-",
      "name": "8BOOL",
      "owners": [],
      "pledge": "1000.00ADA",
      "retired": false,
      "saturation": undefined,
      "size": "- %",
      "ticker": "8BOOL",
      "totalRewards": 10000n,
    },
    "id": "39deffa1dfcfe192ea0efeb3e9bcd9878190627fb590ec81f390cd6d",
    "name": "8BOOL",
    "stakePool": {
      "cost": 340000000n,
      "hexId": "39deffa1dfcfe192ea0efeb3e9bcd9878190627fb590ec81f390cd6d",
      "id": "pool18800lgwlelse96swl6e7n0xes7qeqcnlkkgweq0njrxk66zlj9a",
      "metadata": {
        "name": "8BOOL",
        "ticker": "8BOOL",
      },
      "metrics": {},
      "pledge": 1000000000n,
    },
    "ticker": "8BOOL",
    "value": 1n,
    "weight": 33,
  },
  {
    "displayData": {
      "apy": undefined,
      "cost": "-% + 340ADA",
      "description": undefined,
      "fee": "340.00",
      "hexId": "39faf89aba3daab2bce656a9698b0a2c05e679c0fb360ee6e3b46acb",
      "id": "pool188a03x468k4t908x265knzc29sz7v7wqlvmqaehrk34vkll3e2h",
      "isStakingPool": undefined,
      "lastReward": 2222n,
      "logo": "data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2230%22%20height%3D%2230%22%20viewBox%3D%220%200%2030%2030%22%3E%3Cpath%20fill%3D%22%23329954%22%20d%3D%22M12%203L15%206L12%209L9%206ZM21%206L18%209L15%206L18%203ZM18%2027L15%2024L18%2021L21%2024ZM9%2024L12%2021L15%2024L12%2027ZM6%209L9%2012L6%2015L3%2012ZM27%2012L24%2015L21%2012L24%209ZM24%2021L21%2018L24%2015L27%2018ZM3%2018L6%2015L9%2018L6%2021Z%22%2F%3E%3Cpath%20fill%3D%22%23b2e5c3%22%20d%3D%22M6%209L3%206L6%203L9%206ZM21%206L24%203L27%206L24%209ZM24%2021L27%2024L24%2027L21%2024ZM9%2024L6%2027L3%2024L6%2021Z%22%2F%3E%3Cpath%20fill%3D%22%2366cc87%22%20d%3D%22M11%2011L15%2011L15%2015L11%2015ZM19%2011L19%2015L15%2015L15%2011ZM19%2019L15%2019L15%2015L19%2015ZM11%2019L11%2015L15%2015L15%2019Z%22%2F%3E%3C%2Fsvg%3E",
      "margin": "-",
      "name": "8BOOM",
      "owners": [],
      "pledge": "7000.00ADA",
      "retired": false,
      "saturation": undefined,
      "size": "- %",
      "ticker": "8BOOM",
      "totalRewards": 2222n,
    },
    "id": "39faf89aba3daab2bce656a9698b0a2c05e679c0fb360ee6e3b46acb",
    "name": "8BOOM",
    "stakePool": {
      "cost": 340000000n,
      "hexId": "39faf89aba3daab2bce656a9698b0a2c05e679c0fb360ee6e3b46acb",
      "id": "pool188a03x468k4t908x265knzc29sz7v7wqlvmqaehrk34vkll3e2h",
      "metadata": {
        "name": "8BOOM",
        "ticker": "8BOOM",
      },
      "metrics": {},
      "pledge": 7000000000n,
    },
    "ticker": "8BOOM",
    "value": 1n,
    "weight": 33,
  },
  {
    "displayData": {
      "apy": undefined,
      "cost": "-% + 340ADA",
      "description": undefined,
      "fee": "340.00",
      "hexId": "3867a09729a1f954762eea035a82e2d9d3a14f1fa791a022ef0da242",
      "id": "pool18pn6p9ef58u4ga3wagp44qhzm8f6zncl57g6qgh0pk3yytwz54h",
      "isStakingPool": undefined,
      "lastReward": 3333n,
      "logo": "data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2230%22%20height%3D%2230%22%20viewBox%3D%220%200%2030%2030%22%3E%3Cpath%20fill%3D%22%23e3e3e3%22%20d%3D%22M9%206L12%203L15%206L12%209ZM18%203L21%206L18%209L15%206ZM21%2024L18%2027L15%2024L18%2021ZM12%2027L9%2024L12%2021L15%2024ZM3%2012L6%209L9%2012L6%2015ZM24%209L27%2012L24%2015L21%2012ZM27%2018L24%2021L21%2018L24%2015ZM6%2021L3%2018L6%2015L9%2018Z%22%2F%3E%3Cpath%20fill%3D%22%2359c6c7%22%20d%3D%22M4%206a2%2C2%200%201%2C1%204%2C0a2%2C2%200%201%2C1%20-4%2C0M22%206a2%2C2%200%201%2C1%204%2C0a2%2C2%200%201%2C1%20-4%2C0M22%2024a2%2C2%200%201%2C1%204%2C0a2%2C2%200%201%2C1%20-4%2C0M4%2024a2%2C2%200%201%2C1%204%2C0a2%2C2%200%201%2C1%20-4%2C0M9%209L15%209L15%2015L9%2015ZM11%2014.2L14.2%2014.2L14.2%2011L11%2011ZM21%209L21%2015L15%2015L15%209ZM15.8%2011L15.8%2014.2L19%2014.2L19%2011ZM21%2021L15%2021L15%2015L21%2015ZM19%2015.8L15.8%2015.8L15.8%2019L19%2019ZM9%2021L9%2015L15%2015L15%2021ZM14.2%2019L14.2%2015.8L11%2015.8L11%2019Z%22%2F%3E%3C%2Fsvg%3E",
      "margin": "-",
      "name": "ADA Capital",
      "owners": [],
      "pledge": "1659000.00ADA",
      "retired": false,
      "saturation": undefined,
      "size": "- %",
      "ticker": "ADACT",
      "totalRewards": 3333n,
    },
    "id": "3867a09729a1f954762eea035a82e2d9d3a14f1fa791a022ef0da242",
    "name": "ADA Capital",
    "stakePool": {
      "cost": 340000000n,
      "hexId": "3867a09729a1f954762eea035a82e2d9d3a14f1fa791a022ef0da242",
      "id": "pool18pn6p9ef58u4ga3wagp44qhzm8f6zncl57g6qgh0pk3yytwz54h",
      "metadata": {
        "name": "ADA Capital",
        "ticker": "ADACT",
      },
      "metrics": {},
      "pledge": 1659000000000n,
    },
    "ticker": "ADACT",
    "value": 1n,
    "weight": 33,
  },
]
`;
